File: clients/go/jobs.go
package agenticdriver

import (
	"context"
	"encoding/json"
	"regexp"
	"strings"
	"time"
)

type JobSubmit struct {
	Key     string  `json:"key"`
	Request Request `json:"request"`
}
type JobIdentity struct {
	ID string `json:"id"`
}
type JobEventsRequest struct {
	ID    string `json:"id"`
	After int64  `json:"after"`
	Limit int    `json:"limit,omitempty"`
}
type JobInfo struct {
	ID              string  `json:"id"`
	RunID           string  `json:"runId"`
	Provider        string  `json:"provider"`
	Model           string  `json:"model"`
	State           string  `json:"state"`
	Cursor          int64   `json:"cursor"`
	CancelRequested bool    `json:"cancelRequested"`
	CreatedAt       string  `json:"createdAt"`
	UpdatedAt       string  `json:"updatedAt"`
	ExpiresAt       *string `json:"expiresAt,omitempty"`
}
type JobEventPage struct {
	Job        JobInfo `json:"job"`
	Events     []Event `json:"events"`
	NextCursor int64   `json:"nextCursor"`
	HasMore    bool    `json:"hasMore"`
}

var jobID = regexp.MustCompile(`^[a-f0-9]{8}(?:-[a-f0-9]{4}){3}-[a-f0-9]{12}$`)

func invalidJob() error {
	return &Error{Code: "INVALID_RESPONSE", Message: "Invalid job metadata or event page."}
}
func jobDate(value string) bool {
	_, err := time.Parse(time.RFC3339Nano, value)
	return err == nil && sessionDate.MatchString(value)
}
func jobTerminal(state string) bool { return state != "queued" && state != "running" }
func (info *JobInfo) UnmarshalJSON(data []byte) error {
	type plain JobInfo
	fields, ok := object(data)
	var value plain
	if !ok || json.Unmarshal(exactSessionFields(fields, "id", "runId", "provider", "model", "state", "cursor", "cancelRequested", "createdAt", "updatedAt", "expiresAt"), &value) != nil ||
		!jobID.MatchString(value.ID) || !jobID.MatchString(value.RunID) || !stringField(fields, "provider", false) || !stringField(fields, "model", false) ||
		!numberValue(fields["cursor"], false, true) || !boolValue(fields["cancelRequested"]) || !jobDate(value.CreatedAt) || !jobDate(value.UpdatedAt) || value.UpdatedAt < value.CreatedAt {
		return invalidJob()
	}
	switch value.State {
	case "queued", "running", "completed", "failed", "cancelled", "interrupted":
	default:
		return invalidJob()
	}
	_, expiry := fields["expiresAt"]
	terminal := jobTerminal(value.State)
	if terminal != expiry || (expiry && (value.ExpiresAt == nil || !jobDate(*value.ExpiresAt) || *value.ExpiresAt <= value.UpdatedAt)) ||
		(value.State == "queued" && value.Cursor != 0) || (terminal && value.Cursor < 2) {
		return invalidJob()
	}
	*info = JobInfo(value)
	return nil
}
func (c *Client) SubmitJob(ctx context.Context, request JobSubmit) (JobInfo, error) {
	result, err := c.jobRequest(ctx, "submit", request)
	if err == nil && (result.Provider != request.Request.Provider || result.Model != request.Request.Model) {
		err = invalidJob()
	}
	return result, err
}
func (c *Client) ReadJob(ctx context.Context, request JobIdentity) (JobInfo, error) {
	result, err := c.jobRequest(ctx, "read", request)
	if err == nil && result.ID != request.ID {
		err = invalidJob()
	}
	return result, err
}
func (c *Client) CancelJob(ctx context.Context, request JobIdentity) (JobInfo, error) {
	result, err := c.jobRequest(ctx, "cancel", request)
	if err == nil && result.ID != request.ID {
		err = invalidJob()
	}
	return result, err
}
func (c *Client) jobRequest(ctx context.Context, operation string, input any) (JobInfo, error) {
	var result JobInfo
	res, err := c.request(ctx, "v1/jobs/"+operation, input, false)
	if err != nil {
		return result, err
	}
	err = decode(res, &result)
	return result, err
}

// JobEvents returns observations only; it does not execute tools or retry jobs.
func (c *Client) JobEvents(ctx context.Context, input JobEventsRequest) (JobEventPage, error) {
	var raw json.RawMessage
	var page JobEventPage
	res, err := c.request(ctx, "v1/jobs/events", input, false)
	if err != nil {
		return page, err
	}
	if err = decode(res, &raw); err != nil {
		return page, err
	}
	fields, ok := object(raw)
	var events []json.RawMessage
	if !ok || json.Unmarshal(fields["job"], &page.Job) != nil || json.Unmarshal(fields["events"], &events) != nil || events == nil ||
		!numberValue(fields["nextCursor"], false, true) || json.Unmarshal(fields["nextCursor"], &page.NextCursor) != nil ||
		!boolValue(fields["hasMore"]) || json.Unmarshal(fields["hasMore"], &page.HasMore) != nil || page.Job.ID != input.ID {
		return page, invalidJob()
	}
	limit := input.Limit
	if limit == 0 {
		limit = 100
	}
	if len(events) > limit || len(events) > 100 {
		return page, invalidJob()
	}
	cursor := input.After
	page.Events = []Event{}
	for _, rawEvent := range events {
		fields, ok := object(rawEvent)
		var event Event
		if !ok || json.Unmarshal(exactSessionFields(fields, "type", "runId", "sequence", "timestamp", "optional", "provider", "model", "step", "phase", "call", "callId", "output", "usage", "text", "result", "error"), &event) != nil {
			return page, invalidJob()
		}
		request := Request{Provider: page.Job.Provider, Model: page.Job.Model}
		if event.Result != nil && event.Result.Retrieval != nil {
			request.Retrieval = &RetrievalRequest{Corpus: event.Result.Retrieval.Corpus, Limit: 16, MaxContextBytes: 2000000}
		}
		cursor++
		if !knownEvent(event.Type) || strings.HasPrefix(event.Type, "approval.") || event.Type == "tool.execution.requested" ||
			!eventValid(rawEvent, event, request, cursor == 1) || int64(event.Sequence) != cursor || event.RunID != page.Job.RunID || cursor > page.Job.Cursor {
			return page, invalidJob()
		}
		ended := event.Type == "run.completed" || event.Type == "run.failed" || event.Type == "run.cancelled"
		terminal := jobTerminal(page.Job.State)
		if (ended && (cursor != page.Job.Cursor || !terminal)) || (cursor == page.Job.Cursor && terminal && !ended) ||
			(event.Type == "run.completed" && page.Job.State != "completed") || (event.Type == "run.cancelled" && page.Job.State != "cancelled") ||
			(event.Type == "run.failed" && (page.Job.State != "failed" && page.Job.State != "interrupted" || page.Job.State == "interrupted" && event.Error.Outcome != "uncertain")) {
			return page, invalidJob()
		}
		event.Raw = rawEvent
		page.Events = append(page.Events, event)
	}
	if page.NextCursor != cursor || cursor > page.Job.Cursor || page.HasMore != (cursor < page.Job.Cursor) || (page.HasMore && len(events) == 0) {
		return page, invalidJob()
	}
	return page, nil
}

File: clients/rust/src/jobs.rs
//! Explicit detached jobs. Reading an event page never executes application tools.
use crate::{Event, RunRequest};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize)]
pub struct JobSubmit {
    pub key: String,
    pub request: RunRequest,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobIdentity {
    pub id: String,
}
#[derive(Debug, Serialize, Deserialize)]
pub struct JobEventsRequest {
    pub id: String,
    pub after: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum JobState {
    Queued,
    Running,
    Completed,
    Failed,
    Cancelled,
    Interrupted,
}
impl JobState {
    pub fn is_terminal(self) -> bool {
        !matches!(self, Self::Queued | Self::Running)
    }
}
#[derive(Debug, Deserialize)]
#[serde(try_from = "Value")]
pub struct JobInfo {
    pub id: String,
    pub run_id: String,
    pub provider: String,
    pub model: String,
    pub state: JobState,
    pub cursor: u64,
    pub cancel_requested: bool,
    pub created_at: String,
    pub updated_at: String,
    pub expires_at: Option<String>,
}
impl JobInfo {
    pub fn identity(&self) -> JobIdentity {
        JobIdentity {
            id: self.id.clone(),
        }
    }
}
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JobEventPage {
    pub job: JobInfo,
    pub events: Vec<Event>,
    pub next_cursor: u64,
    pub has_more: bool,
}
fn id_valid(value: &str) -> bool {
    value.len() == 36
        && value.bytes().enumerate().all(|(i, b)| {
            if [8, 13, 18, 23].contains(&i) {
                b == b'-'
            } else {
                b.is_ascii_digit() || (b'a'..=b'f').contains(&b)
            }
        })
}
fn date_valid(value: &str) -> bool {
    value.len() == 24
        && value.as_bytes()[19] == b'.'
        && value.ends_with('Z')
        && crate::validation::timestamp_valid(value)
}
impl TryFrom<Value> for JobInfo {
    type Error = &'static str;
    fn try_from(value: Value) -> std::result::Result<Self, Self::Error> {
        let invalid = "Invalid durable job metadata";
        let text = |key: &str| value.get(key).and_then(Value::as_str).unwrap_or("");
        let state: JobState = serde_json::from_value(value.get("state").cloned().ok_or(invalid)?)
            .map_err(|_| invalid)?;
        let cursor = value.get("cursor").and_then(Value::as_u64).ok_or(invalid)?;
        let cancel_requested = value
            .get("cancelRequested")
            .and_then(Value::as_bool)
            .ok_or(invalid)?;
        if !id_valid(text("id"))
            || !id_valid(text("runId"))
            || text("provider").is_empty()
            || text("model").is_empty()
            || cursor > 9_007_199_254_740_991
            || !date_valid(text("createdAt"))
            || !date_valid(text("updatedAt"))
            || text("updatedAt") < text("createdAt")
            || state.is_terminal() != value.get("expiresAt").is_some()
            || (state.is_terminal()
                && (!date_valid(text("expiresAt"))
                    || text("expiresAt") <= text("updatedAt")
                    || cursor < 2))
            || (state == JobState::Queued && cursor != 0)
        {
            return Err(invalid);
        }
        Ok(Self {
            id: text("id").into(),
            run_id: text("runId").into(),
            provider: text("provider").into(),
            model: text("model").into(),
            state,
            cursor,
            cancel_requested,
            created_at: text("createdAt").into(),
            updated_at: text("updatedAt").into(),
            expires_at: value.get("expiresAt").map(|_| text("expiresAt").into()),
        })
    }
}
#[cfg(any(feature = "blocking", feature = "async"))]
fn invalid() -> crate::Error {
    crate::protocol_error(
        "INVALID_RESPONSE",
        "Invalid or mismatched durable job response.",
    )
}
#[cfg(any(feature = "blocking", feature = "async"))]
pub(crate) fn info(
    value: Value,
    submit: Option<&JobSubmit>,
    identity: Option<&JobIdentity>,
) -> crate::Result<JobInfo> {
    let result: JobInfo = serde_json::from_value(value).map_err(|_| invalid())?;
    if submit
        .is_some_and(|s| result.provider != s.request.provider || result.model != s.request.model)
        || identity.is_some_and(|i| result.id != i.id)
    {
        return Err(invalid());
    }
    Ok(result)
}
#[cfg(any(feature = "blocking", feature = "async"))]
pub(crate) fn page(value: Value, input: &JobEventsRequest) -> crate::Result<JobEventPage> {
    let page: JobEventPage = serde_json::from_value(value).map_err(|_| invalid())?;
    if page.job.id != input.id
        || page.events.len() > input.limit.unwrap_or(100).min(100) as usize
        || input.after > 9_007_199_254_740_991
    {
        return Err(invalid());
    }
    let mut cursor = input.after;
    for event in &page.events {
        let mut request = RunRequest::new(&page.job.provider, &page.job.model, "");
        if let Some(retrieval) = event.result.as_ref().and_then(|r| r.retrieval.as_ref()) {
            request.retrieval = Some(crate::RetrievalRequest {
                corpus: retrieval.corpus.clone(),
                limit: Some(16),
                max_context_bytes: Some(2_000_000),
                ..Default::default()
            });
        }
        cursor += 1;
        if !matches!(
            event.kind.as_str(),
            "run.started"
                | "step.started"
                | "text.delta"
                | "run.progress"
                | "tool.called"
                | "tool.completed"
                | "usage.reported"
                | "run.completed"
                | "run.failed"
                | "run.cancelled"
        ) || !crate::validation::event_valid(event, &request, cursor == 1)
            || event.sequence != cursor
            || event.run_id != page.job.run_id
            || cursor > page.job.cursor
        {
            return Err(invalid());
        }
        let ended = event.is_terminal();
        if (ended && (cursor != page.job.cursor || !page.job.state.is_terminal()))
            || (cursor == page.job.cursor && page.job.state.is_terminal() && !ended)
            || (event.kind == "run.completed" && page.job.state != JobState::Completed)
            || (event.kind == "run.cancelled" && page.job.state != JobState::Cancelled)
            || (event.kind == "run.failed"
                && (!matches!(page.job.state, JobState::Failed | JobState::Interrupted)
                    || (page.job.state == JobState::Interrupted
                        && event.error.as_ref().and_then(|e| e.outcome)
                            != Some(crate::ErrorOutcome::Uncertain))))
        {
            return Err(invalid());
        }
    }
    if page.next_cursor != cursor
        || cursor > page.job.cursor
        || page.has_more != (cursor < page.job.cursor)
        || (page.has_more && page.events.is_empty())
    {
        return Err(invalid());
    }
    Ok(page)
}
